File: crates/1bit-helm-tui/src/pane.rs
//! Pane tree — recursive splits with per-node ratio.
//!
//! Skeleton only; the v1 layout rendered by `widgets::draw_root_panes`
//! hardcodes three panes (status / logs / repl). This type will replace
//! that once resize + rearrange bindings land.

#![allow(dead_code)]

use serde::{Deserialize, Serialize};

/// Root of the pane tree. `serde` round-trips through the layout JSON
/// at `~/.config/1bit/tui-layout.json`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum Node {
    /// Leaf — renders one named widget.
    Leaf {
        /// Widget key — must match one of [`crate::widgets::WIDGET_KEYS`].
        widget: String,
    },
    /// Vertical split — two children stacked top/bottom.
    VerticalSplit {
        /// 0.0–1.0 top child height ratio. Default 0.5.
        ratio: f32,
        /// Top child.
        top: Box<Node>,
        /// Bottom child.
        bottom: Box<Node>,
    },
    /// Horizontal split — two children side by side.
    HorizontalSplit {
        /// 0.0–1.0 left child width ratio. Default 0.5.
        ratio: f32,
        /// Left child.
        left: Box<Node>,
        /// Right child.
        right: Box<Node>,
    },
}

impl Node {
    /// Built-in default: status pane top-left, gpu top-right, logs full-bottom.
    pub fn default_layout() -> Self {
        Node::VerticalSplit {
            ratio: 0.65,
            top: Box::new(Node::HorizontalSplit {
                ratio: 0.60,
                left: Box::new(Node::Leaf { widget: "status".into() }),
                right: Box::new(Node::Leaf { widget: "gpu".into() }),
            }),
            bottom: Box::new(Node::Leaf { widget: "logs".into() }),
        }
    }

    /// Count leaf panes recursively. Used by the pane-cycle binding.
    pub fn leaf_count(&self) -> usize {
        match self {
            Node::Leaf { .. } => 1,
            Node::VerticalSplit { top, bottom, .. } => top.leaf_count() + bottom.leaf_count(),
            Node::HorizontalSplit { left, right, .. } => left.leaf_count() + right.leaf_count(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_layout_has_three_leaves() {
        assert_eq!(Node::default_layout().leaf_count(), 3);
    }

    #[test]
    fn round_trip_default_layout_json() {
        let n = Node::default_layout();
        let j = serde_json::to_string(&n).unwrap();
        let n2: Node = serde_json::from_str(&j).unwrap();
        assert_eq!(n.leaf_count(), n2.leaf_count());
    }
}

File: cpp/helm-tui/src/pane.cpp
#include "onebit/helm_tui/pane.hpp"

#include <nlohmann/json.hpp>

#include <stdexcept>
#include <utility>

namespace onebit::helm_tui {

Node Node::leaf(std::string widget)
{
    return Node{LeafNode{std::move(widget)}};
}

Node Node::vsplit(float ratio, Node top, Node bottom)
{
    return Node{VerticalSplit{
        ratio,
        std::make_unique<Node>(std::move(top)),
        std::make_unique<Node>(std::move(bottom)),
    }};
}

Node Node::hsplit(float ratio, Node left, Node right)
{
    return Node{HorizontalSplit{
        ratio,
        std::make_unique<Node>(std::move(left)),
        std::make_unique<Node>(std::move(right)),
    }};
}

Node Node::default_layout()
{
    return vsplit(
        0.65F,
        hsplit(0.60F,
               leaf("status"),
               leaf("gpu")),
        leaf("logs"));
}

std::size_t Node::leaf_count() const noexcept
{
    if (const auto* l = std::get_if<LeafNode>(&v_)) {
        (void)l;
        return 1;
    }
    if (const auto* vs = std::get_if<VerticalSplit>(&v_)) {
        return vs->top->leaf_count() + vs->bottom->leaf_count();
    }
    if (const auto* hs = std::get_if<HorizontalSplit>(&v_)) {
        return hs->left->leaf_count() + hs->right->leaf_count();
    }
    return 0;
}

nlohmann::json node_to_json(const Node& n)
{
    using nlohmann::json;
    return std::visit(
        [](const auto& alt) -> json {
            using T = std::decay_t<decltype(alt)>;
            if constexpr (std::is_same_v<T, LeafNode>) {
                return json{{"kind", "leaf"}, {"widget", alt.widget}};
            } else if constexpr (std::is_same_v<T, VerticalSplit>) {
                return json{{"kind", "vertical_split"},
                            {"ratio", alt.ratio},
                            {"top", node_to_json(*alt.top)},
                            {"bottom", node_to_json(*alt.bottom)}};
            } else {
                return json{{"kind", "horizontal_split"},
                            {"ratio", alt.ratio},
                            {"left", node_to_json(*alt.left)},
                            {"right", node_to_json(*alt.right)}};
            }
        },
        n.variant());
}

Node node_from_json(const nlohmann::json& j)
{
    if (!j.is_object() || !j.contains("kind") || !j["kind"].is_string()) {
        throw std::runtime_error("layout: missing 'kind' tag");
    }
    const std::string kind = j["kind"].get<std::string>();
    if (kind == "leaf") {
        return Node::leaf(j.value("widget", std::string{}));
    }
    if (kind == "vertical_split") {
        return Node::vsplit(j.value("ratio", 0.5F),
                            node_from_json(j.at("top")),
                            node_from_json(j.at("bottom")));
    }
    if (kind == "horizontal_split") {
        return Node::hsplit(j.value("ratio", 0.5F),
                            node_from_json(j.at("left")),
                            node_from_json(j.at("right")));
    }
    throw std::runtime_error("layout: unknown kind '" + kind + "'");
}

} // namespace onebit::helm_tui
